Add ability to set minion ID by module

skizunov opened the pull request:
### What does this PR do?

Add a new configuration option `minion_id_type`. Solution
based on configuration option `master_type`.
`eval_minion_id_func` based on `salt.minion.eval_master_func`.


### What issues does this PR fix or reference?

Fixes #15389.

### Tests written?

No

skizunov committed:
Add ability to set minion ID by module

Fixes #15389. Add a new configuration option `minion_id_type`. Solution
based on configuration option `master_type`.
`eval_minion_id_func` based on `salt.minion.eval_master_func`.

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

Conflicts:
	salt/config/__init__.py

salt-jenkins commented:
@skizunov, thanks for your PR! By analyzing the history of the files in this pull request, we identified @basepi, @isbm and @ahammond to be potential reviewers.

isbm commented on salt/config/__init__.py:
```diff
@@ -840,6 +840,13 @@ def _gather_buffer_space():
+    # the minion ID or left unspecified for default minion ID generation
+    # behavior. If 'func' is specified, the 'id' option should be set to an
+    # exec module function to run to determine the minion ID.
+    'minion_id_type': str,
```
This is confusing. If that supposed to be a function from a module, then I'd call it as `minion_id_function` or something. And the interface to the user would be something like:
```yaml
minion_id_function: salt.modules.mymodule.myfunction
```
In the example above you would not need to have two options in config: this one and `id` as an actual function. An absence of this parameter would imply `str` and thus generate an id as usual.

skizunov replied:
I did consider this approach. Given that `master_type` is already implemented, I decided to follow the approach that is already in use (and hence supposedly the safer bet for a design choice in a new feature). If someone else agrees with you to use your approach instead, I am willing to change it.
@cachedout : Do you have an opinion on this?

cachedout replied:
It might take me a day or two to get to this. I will prioritize it though. Apologies for the delay.

isbm replied:
@skizunov `master_type` is something else in this case and has more than two state. In your case it is literally _one_ state: it is either external, or nothing. So why complicate this?

skizunov replied:
One thought is that it gives us the ability to easily extend this in the future. In case we want to get it from yet another source (file?)

isbm replied:
Uggh, I'd not go there, because [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it). If one needs to take something from a file, then just:
```python
def foobar(*args, **kwargs):
    return open(kwargs.get('filename')).read()
```
And then your config would be:
```yaml
minion_id_function:
    module.foobar:
        filename: /etc/hostname
```
As an example.

cachedout replied:
The notion that somebody should have to write a function just to use a file to read a minion id strikes me as unnecessary indirection. The way that @skizunov has documented it makes it quite clear what should be done in order to achieve the desired functionality and leaves the door open for additional options.

isbm replied:
@cachedout Hmm. I would rather go here in "general mechanism", like if you're need to first run some 3rd party Perl script that creates a file that you  will read and a bit parse — then you _do need_ a function. @skizunov only described a simple thing that will end-up an appendix code to the other things. And that part I don't like. If we are doing pluggable ID generator, let it be overall pluggable, even if this is just to read a plain text from somewhere.

isbm commented on salt/config/__init__.py:
```diff
@@ -3284,6 +3292,42 @@ def _cache_id(minion_id, cache_file):
+    if '__minion_id_func_evaluated' in opts:
+        return opts['id']
+
+    import salt.loader as loader
```
This belongs above in the imports.

skizunov replied:
This was put here purposely to avoid this import in other cases. Is there a specific reason why it should be above in the imports, or this this just pedantic style enforcement?

isbm replied:
That would make sense if it would be something very unique and long-loading memory consuming. But it is already imported elsewhere. Python doesn't acquire some more memory specifically for it, just creates a ref to it, which is basically nothing. So I'd avoid scattered imports.

skizunov replied:
I don't believe we can move this import to the top. The reason is that `salt/loader.py` imports `salt.config` near the top of the file. If we add this import to the top of `salt/config/__init__.py` we would be creating a circular dependency which would break things. A comment explaining this is probably in order though.

cachedout replied:
I believe that @skizunov is correct here.

isbm commented on salt/config/__init__.py:
```diff
@@ -3284,6 +3292,42 @@ def _cache_id(minion_id, cache_file):
+    try:
+        minion_id_mod = loader.raw_mod(opts, mod, fun)
+        if not minion_id_mod:
+            raise KeyError
```
I'd expect here output to the error log with a clear explanation what is happening.

skizunov replied:
Look below and you will see something catching the exception and logging it. This is based on the design of salt.minion.eval_master_func which does something similar.

isbm commented on salt/config/__init__.py:
```diff
@@ -3284,6 +3292,42 @@ def _cache_id(minion_id, cache_file):
+        if not minion_id_mod:
+            raise KeyError
+        # we take whatever the module returns as the minion ID
+        newid = minion_id_mod[mod_fun]()
```
No input interface from anywhere? There is a plenty of data in pillars, why not getting some?

skizunov replied:
I am confused by this comment. Isn't pillar data received from the master, and the minion_id needs to be generated before connection to the master?

isbm replied:
@skizunov hint: salt-ssh and `--local`.

cachedout replied:
I am also confused by this. We'll need more than just a hint here in order to discuss what you believe the underlying issue to be. :]

isbm replied:
@cachedout Salt SSH has no `--local` option, sorry. This is `salt-call` does. But the whole thing is this: imagine you're bootstrapping a machine that has only SSH access. Just like we do with our SUSE Manager :wink: Now, you're supplying some data along, say your function normally creates nice IDs according to the "DNS name + Datacenter prefix", as for an example. In one DC we end up `foobar-muenchen.foo.com` and another `foobar-frankfurt.foo.com` But then where this `$PREFIX` comes from? So I am wondering if that could be considered along.

skizunov replied:
@isbm : The feature I added is very powerful. It will call into any module function that you write. Not only that, but that module function will have access to grains. So within that function, you could easily implement the policy you mentioned above. In your case, `$PREFIX` would come from one of the networking grains, or another option is you can make any Python call you like to generate it.

isbm replied:
@skizunov OK, so what you say is that the parametrization is going to the function alone, not to the applier. Fair enough. We still have design questions here though (the other threads).

isbm commented on salt/config/__init__.py:
```diff
@@ -3284,6 +3292,42 @@ def _cache_id(minion_id, cache_file):
+                newid, mod_fun)
+            )
+            sys.exit(salt.defaults.exitcodes.EX_GENERIC)
+        opts['__minion_id_func_evaluated'] = True
```
I find this option unnecessary. Just specify a module in the option earlier mentioned, and keep `id` to none, which will your indication to evaluate, or not. KISS.

isbm commented on salt/config/__init__.py:
```diff
@@ -3325,13 +3369,19 @@ def get_id(opts, cache_minion_id=False):
+    if opts.get('minion_id_type', 'str') == 'func':
+        newid = eval_minion_id_func(opts)
+    else:
+        newid = salt.utils.network.generate_minion_id()
```
I'd not design it this way. The eval_minion_id_func should just fall-back inside. So you won't need to check this numerous times here/there, but just do it once and for all:
```python
newid = eval_minion_id_func(opts)
```
And that will do `generate_minion_id()` inside in case opts aren't `func`. But again, I would not use this `func`/`type` thing, but rather just a module name directly. So my code would be like:

```python
custom_module = opts.get('minion_id_function')
newid = minion_id_mod[custom_module](*args, **kwargs) if custom_module \
            else salt.utils.network.generate_minion_id()
```

isbm commented on salt/config/__init__.py:
```diff
@@ -3325,13 +3369,19 @@ def get_id(opts, cache_minion_id=False):
+        if opts.get('minion_id_type', 'str') == 'func':
+            log.debug('Found minion id from eval_minion_id_func(): {0}'.format(newid))
+        else:
+            log.debug('Found minion id from generate_minion_id(): {0}'.format(newid))
```
```python
log.debug('Found minion id from {0}(): {1}'.format(
    opts.get('minion_id_type) == 'func' and 'eval_minion_id_func' or 'generate_minion_id', newid)

isbm commented on salt/config/__init__.py:
```diff
@@ -3563,7 +3613,7 @@ def apply_master_config(overrides=None, defaults=None):
     using_ip_for_id = False
     append_master = False
-    if not opts.get('id'):
+    if not opts.get('id') or opts.get('minion_id_type', 'str') == 'func':
```
This is valid to the identical line above. I am not sure this is correct. You are asking to run `get_id()` function if there is no parameter `id` or options set to `func`? Isn't it a bug, because you're supposed to ask it when `minion_id_type` is "str" instead?

And in case you want to `get_id` as usual with `minion_id_type: str` and `id` absent, then it is enough to just do this:
```python
if not opts.get('id', opts.get('minion_id_type' == 'str')):
   ...
```

skizunov replied:
@isbm : I'll just wait for @cachedout to comment on the design and approach before modifying things. This way, I will know if the fundamental design should be changed.

isbm replied:
@skizunov Agreed. But please keep in mind that unit tests are always welcome. :wink:

isbm reviewed (changes requested):
@skizunov please look at these changes, as well as please better document the feature, because it is unknown how it is supposed to be used.

@cachedout I'd rather additionally ask for unit tests here. Too much changes that can break everything.

skizunov commented:
@cachedout : I wanted to ping you on this. I would like your view as to which configuration method is more ideal. The options are:

1. Current method of using `minion_id_type` and then referring to `id` for the function to call when the type is `func` (modeled after `master_type`).
2. The suggested alternative of using `minion_id_function` which would contain the function to call or left empty if the feature is not used.

I would be happy with either approach and changing it would be very little effort. I just want to know which approach you think is the most appropriate so that progress may be made on this issue.

skizunov committed:
Changes from feedback

- Add comment as to why we import inside the function
- Have one log.debug that prints out the appropriate function string

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

skizunov commented:
@cachedout : With the last commit, I believe I addressed most of the open issues. From my understanding, the only open issues remaining are:

1. What @isbm means by `hint: salt-ssh and --local`
2. If a unit test is required for this feature. I couldn't find any unit tests for `master_type`. If a unit test is required, your advice on where to put it would be appreciated.

isbm commented:
@skizunov 

1. What I meant is that it still would be nice to be configurable. Where all that comes is another story. Say, from grains.
2. You can always create your new one! (Unit test)

cachedout commented:
@skizunov Did the clarification from @isbm make sense and did you want a chance to respond?

skizunov commented:
@isbm : For your first point, it is already configurable in 2 ways:
1) You would set which function is to be called as an option (in a configuration file)
2) If you want more granular control, the function called has access to grains and opts (as `__grains__` and `__opts__`), so it can have custom logic to do whatever it wants from there. It could behave differently based on whether it is called as salt-ssh, salt-minion, salt-master, etc. You control the implementation of the function so you can do just about anything you want.

Wouldn't this be sufficient for almost anyone's needs? If you have an alternative suggestion, please be specific.

isbm commented:
@skizunov for 1) and 2) I am fine. But I still am not fine with the way you define it in the config. I still believe that if you define as I already showed before, is still better and more elegant, rather then go define here, there, type `str`, type `func` etc etc. I would like to have something this way:
```yaml
minion_id_function: 
    salt.modules.mymodule.myfunction:
       - some
       - static
       - optional
       - parameters
```
Or just as simple as:
```yaml
minion_id_function: salt.modules.mymodule.myfunction:
```

And how about Unit Test(s)? :wink:

cachedout commented:
I did comment on this, in-line. I wrote `The way that @skizunov has documented it makes it quite clear what should be done in order to achieve the desired functionality and leaves the door open for additional options.`

I will have @thatch45 come in and make a final decision on this, since it doesn't seem likely that there is going to be much agreement here otherwise.

thatch45 commented:
I am going to complicate this.
You both have good arguments, to sum up:
1. Conflating the id field to also be a function worries me, I would like this to be separate from the id option
2. While opts and grains are available to the function, since it is an execution module function and those typically accept input as args and kwargs that facility should also exist, but one should be able to enter just a function to call in a simple way
3. Special cases that fall back to modular systems have proven to be problematic, as an example I would prefer it if the default pillar was just another ext_pillar module. So I would propose that the func call is the default, but the func to call defaults to the classic approach to generate the id.

So I would propose this:
I would say to turn the code that generates the id into an execution module func and make it the default param to an option called `id_func`, then have another option called `id_args` that, if empty, passes no args.
Then you ALWAYS use the modular system, no special default, and the default is a module
So the default function is just set in the `id_func` option in config
Finally that function will only be called if the `id` option is not set, so that the original behavior of the `id` option remains unchanged.

This way existing options don't get conflated, we increase overall modularity(something I am big fan of) and the configuration remains simple since you would only need to add the one option to set the id function.

For the record @skizunov I LOVE this feature addition, and I really like your approach, but I do have to concede that @isbm does present some valid points, but I think we can make an interface that is cleaner and easier to maintain overall.

isbm commented:
@thatch45 , can't agree more. But. :wink: My tweaks to your proposal is that the `id_func` maybe better called `id_function` (it is easier to remember and avoid like _"What was it?? `id_fun`? `id_func`?.."_ etc  **and** have none of `id_args` parameter. Example:

```yaml
# Simple
id_function: somemodule.somefunction

# With some arguments
id_function:
  somemodule.somefunction:
    - look: no_need
    - of_extra: id_args

# Different purposes:
id_function:
  failover:
    somemodule.somefunction:
      - some
      - args
      - and_some: keywords
  chain:
    somemodule.somefunction:
      - on_return: 'localhost', 127.0.0.1
```
To the defence of the proposed interface above, it:

- does comply with the Salt YAML config
- is cleanest and simplest way (unless I am missing something to possibly throw away even more)
- whilst probably in good five nines you ain't gonna need it, but `somemodule.somefunction` can be also an element of a list and thus `id_function` can have several ID functions (for fall-backs? different realms in datacenter? different cloud providers? etc).
- since you _always_ use the modular mechanism (ant this is totally true), you can also have a flags, like the latest section. E.g. "failover" or "chain". For example, "failover" would only work if the default return `None` or crash or something that direction. The "chain" would _always_ pipe through it the previous result and verify if the output is correct. In the case above, if default ID crashes and failover still returns 'localhost', then the last one detects it and does something. But again, the flags thing can be only a possibility, once needed, if we take the configuration approach above.

Does it make sense?

cachedout commented:
@thatch45 Bump on above comment.

thatch45 commented:
Sorry, I missed the reply, yes @isbm , I would be good with that proposal

skizunov committed:
Add config option `id_function`

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

skizunov commented:
@cachedout, @isbm : I implemented this with the `id_function` configuration option similar to how @isbm described it. I have also added unit tests. One notable difference to what @isbm described is that I could not use positional (no name provided) function arguments. This is because it would be ambiguous if an argument is named or positional since function arguments could be dictionaries and Python is not a typed language. I believe Reactor SLS files have the same limitation (only accept named function arguments), so this should be acceptable. This will work with both arg (no default value) and kwarg function arguments since you can name both when calling into a function.

isbm commented:
@skizunov yeah, positional args aren't the best here anyway, because they do not describe what exactly it is, since this is, after all, still a configuration, just nested one. So yeah, I would still vote for clear _readable, self-explainable configuration_, such as:
```yaml
my_great_func:
    name: Vasya
    second_name: Pupkin
    meaning_of_life: 42
```
So no worries here, I'd say. It is still not SLS, but just a _configuration_. :wink:

isbm commented on salt/config/__init__.py:
```diff
@@ -3360,13 +3422,18 @@ def get_id(opts, cache_minion_id=False):
+    if opts.get('id_function'):
+        newid = eval_id_function(opts)
+    else:
+        newid = salt.utils.network.generate_minion_id()
```
Hmm. I would not do this, as here is a potential bug: in your `eval_id_function` you can possibly [return an empty value](https://github.com/saltstack/salt/pull/41619/files#diff-0ddbe12a41fee1a8e552b35086f918ddR3339) and then we have a trouble. Instead, I would do something like:
```python

def find_minion_id(opts):
    '''
    This is your renamed "eval_id_function" that calls 
    salt.utils.network.generate_minion_id() if no value.
    '''
    ...
    # All your "eval_id_function" code, but do not return anywhere.
    ...
    return opts.get('id') or salt.utils.network.generate_minion_id()

...
def get_id(opts, cache_minion_id=False):
    new_id = find_minion_id(opts)  # let the function alone figure out
                                   # required params from opts
    ...
```

skizunov replied:
I fixed it so that if the function returns an empty string, it will hard fail instead of failover to FQDN. I believe this is the preferred behavior because a user is using the `id_function` feature specifically because they want to avoid the default FQDN behavior. As a user of this feature, that is definitely the behavior I would prefer because I don't want FQDN IDs getting to the master, since this is not what is expected and would cause problems based on how my master is set up.

isbm replied:
No hard opinions here.

isbm commented on salt/config/__init__.py:
```diff
@@ -3319,6 +3330,57 @@ def _cache_id(minion_id, cache_file):
+        if fun_kwargs is None:
+            fun_kwargs = {}
+    else:
+        log.error('\'id_function\' option is not a string nor a dictionary')
```
This is a programmer's message that is known only to a programmer. What exactly actions should take admin  in order to fix that? It is pretty unclear what does that mean "not a string nor a dictionary". BTW, in English it would be "nor a string neither a dictionary".

isbm commented on salt/config/__init__.py:
```diff
@@ -3319,6 +3330,57 @@ def _cache_id(minion_id, cache_file):
+    try:
+        id_mod = loader.raw_mod(opts, mod, fun)
+        if not id_mod:
+            raise KeyError
```
I would rather not do it, but just spit something like this:
```python
log.error('Function "foo.bar" could not be loaded because of "{0}", falling back to default'.format(
    exception_message))`.
```

skizunov replied:
The `KeyError` would be caught below and a hard error would occur. I believe this is the preferred behavior because a user is using the `id_function` feature specifically because they want to avoid the default FQDN behavior.

isbm replied:
Right. OK.

isbm commented on salt/config/__init__.py:
```diff
@@ -3319,6 +3330,57 @@ def _cache_id(minion_id, cache_file):
+        # we take whatever the module returns as the minion ID
+        newid = id_mod[mod_fun](**fun_kwargs)
+        if not isinstance(newid, str):
+            log.error('{0} returned from {1} is not a string'.format(
```
Can we rephrase this? Something like: 
```python
log.error('Function {0} returned value "{1}" of type {2} instead of string'.format(
    mod_fun, new_id, type(new_id)))
```

isbm commented on salt/config/__init__.py:
```diff
@@ -3319,6 +3330,57 @@ def _cache_id(minion_id, cache_file):
+            )
+            sys.exit(salt.defaults.exitcodes.EX_GENERIC)
+        log.info('Evaluated minion ID from module: {0}'.format(mod_fun))
+        return newid
```
I'd _suggest_ to have this return at the end and only there. Just to have less spaghetti over time, when this will get changed/improved.

skizunov replied:
I am going to leave it as is. I don't think there is anything abnormal about having a `return` statement at the end of the inside of a `try` block.

isbm commented on salt/config/__init__.py:
```diff
@@ -3360,13 +3422,18 @@ def get_id(opts, cache_minion_id=False):
-        log.debug('Found minion id from generate_minion_id(): {0}'.format(newid))
+        log.debug('Found minion id from {0}(): {1}'.format(
+            opts.get('id_function') and 'eval_id_function' or
+            'generate_minion_id', newid))
```
The `id_function` name is lost here, while is important. I'd include it as well.

isbm commented on tests/unit/config/test_config.py:
```diff
@@ -447,6 +447,30 @@ def test_master_pillar_roots_glob(self):
+    def test_master_id_function(self):
+        try:
+            tempdir = tempfile.mkdtemp(dir=TMP)
+            master_config = os.path.join(tempdir, 'master')
```
I know you've just copied this from other tests, but normally in the unit tests we are always mocking the file descriptors instead of writing actual files _nowadays_. Because this is not always true that your unit test env has a write access to the temporary system directory, it can be a limited resources sandbox in a read-only docker, for example. Unit tests supposed to run everywhere without touching actual things. Integration tests actually do writing data. Therefore please mock this one instead.

skizunov replied:
All the other test cases in this file do it the way I have done it. For the sake of consistency, I will leave it the way it is. If you believe that this approach is not the best approach in general, perhaps a separate PR that changes all of the test cases is a better approach. Changing the way the tests are executed in this file is beyond the scope of this PR.

isbm replied:
@skizunov :+1:

isbm commented on tests/unit/config/test_config.py:
```diff
@@ -507,6 +531,29 @@ def test_minion_pillar_roots_glob(self):
+            tempdir = tempfile.mkdtemp(dir=TMP)
+            minion_config = os.path.join(tempdir, 'minion')
+
+            with salt.utils.fopen(minion_config, 'w') as fp_:
```
Same here. The file descriptor has to be mocked in-memory instead.

skizunov committed:
Updates due to feedback

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

skizunov commented on salt/config/__init__.py:
```diff
@@ -3319,6 +3330,57 @@ def _cache_id(minion_id, cache_file):
         log.error('Could not cache minion ID: {0}'.format(exc))
 
 
+def eval_id_function(opts):
```
I went with `call_id_function`. `call_external_id_gen_function` is very descriptive, but was a bit long-winded for my tastes.

skizunov commented:
@cachedout : I believe there are no more open issues from @isbm. So whenever you start merging general PRs again, there shouldn't be anything blocking this PR unless you would like me to do something more.
